core/src/main/java/org/apache/accumulo/core/spi/balancer/DefaultLoadBalancer.java

```diff
@@ -0,0 +1,372 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.spi.balancer;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.accumulo.core.client.AccumuloException;
+import org.apache.accumulo.core.client.AccumuloSecurityException;
+import org.apache.accumulo.core.data.TableId;
+import org.apache.accumulo.core.data.TabletId;
+import org.apache.accumulo.core.manager.balancer.TabletServerIdImpl;
+import org.apache.accumulo.core.spi.balancer.data.TServerStatus;
+import org.apache.accumulo.core.spi.balancer.data.TableStatistics;
+import org.apache.accumulo.core.spi.balancer.data.TabletMigration;
+import org.apache.accumulo.core.spi.balancer.data.TabletServerId;
+import org.apache.accumulo.core.spi.balancer.data.TabletStatistics;
+import org.apache.accumulo.core.spi.balancer.util.ThrottledBalancerProblemReporter;
+import org.apache.accumulo.core.spi.balancer.util.ThrottledBalancerProblemReporter.OutstandingMigrationsProblem;
+import org.apache.accumulo.core.spi.balancer.util.ThrottledBalancerProblemReporter.Problem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @since 2.1.0
+ */
+public class DefaultLoadBalancer implements TabletBalancer {
```

Comment:
I wonder if the interface should be called `TabletLoadBalancer` instead of `TabletBalancer`. I think it might make things more clear.

For the default implementation, it might be nice to name it something based on what it does, rather than it having been set in the default config. That way, if the default changes in future, we don't have a class called `DefaultLoadBalancer` that is not actually configured as the default implementation. (Consider, for example, the "default Java GC" implementations, which have specific names and the selected default has changed over time.)

Reply:
@ctubbsii I agree that these changes make sense. I wasn't really trying to improve the interface in order to limit the scope of the work and the review for people. I can certainly make these changes now if you prefer, though. Or, I can add a ticket and submit changes like this (and others, such as: Keith and I also talked about moving the output parameters in the getAssignments and balance calls to return objects instead) in a follow-in PR. The only gotcha there is we need to be sure it's done before 2.1 or else we'll end up having to deprecate code we just added, which is clearly not a good idea.

Reply:
From the perspective of someone reviewing, I like the idea of improving the new SPI in follow on PRs/issues.  Will be easier to review.

> The only gotcha there is we need to be sure it's done before 2.1 or else we'll end up having to deprecate code we just added, which is clearly not a good idea.

Can use blocker label on issues to make sure they are not forgotten for 2.1.   There is still the risk that its never done.  Even if its never improved, I think these new SPIs are a big improvement over the current state.